File: jwt/src/main/java/com/member/jwt/config/SecurityConfig.java
package com.member.jwt.config;

import com.member.jwt.jwt.JWTFilter;
import com.member.jwt.jwt.JWTUtil;
import com.member.jwt.repository.MemberRepository;
import com.member.jwt.repository.RefreshRepository;
import com.member.jwt.service.CustomMemberDetailService;
import com.member.jwt.service.TokenBlacklistService;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.dao.DaoAuthenticationProvider;
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;

import java.util.Collections;

@Configuration
@EnableWebSecurity
public class SecurityConfig {

    private final AuthenticationConfiguration authenticationConfiguration;
    private final JWTUtil jwtUtil;
    private final RefreshRepository refreshRepository;
    private final CustomMemberDetailService customMemberDetailService;
    private final TokenBlacklistService tokenBlacklistService;

    public SecurityConfig(AuthenticationConfiguration authenticationConfiguration, JWTUtil jwtUtil, RefreshRepository refreshRepository, CustomMemberDetailService customMemberDetailService, TokenBlacklistService tokenBlacklistService) {
        this.authenticationConfiguration = authenticationConfiguration;
        this.jwtUtil = jwtUtil;
        this.refreshRepository = refreshRepository;
        this.customMemberDetailService = customMemberDetailService;
        this.tokenBlacklistService = tokenBlacklistService;
    }

    @Bean
    public BCryptPasswordEncoder bCryptPasswordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    public DaoAuthenticationProvider authenticationProvider() {
        DaoAuthenticationProvider authProvider = new DaoAuthenticationProvider();
        authProvider.setUserDetailsService(customMemberDetailService);
        authProvider.setPasswordEncoder(bCryptPasswordEncoder());
        return authProvider;
    }

    @Bean
    public AuthenticationManager authenticationManager() throws Exception {
        AuthenticationManager authenticationManager = authenticationConfiguration.getAuthenticationManager();
        ((org.springframework.security.authentication.ProviderManager) authenticationManager).getProviders().add(authenticationProvider());
        return authenticationManager;
    }

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http, RefreshRepository refreshRepository, MemberRepository memberRepository) throws Exception {
        http
                .csrf((csrf) -> csrf.disable())
                .formLogin((form) -> form.disable())
                .httpBasic((basic) -> basic.disable())
                .authorizeHttpRequests((auth) -> auth
                        .requestMatchers("/login", "/", "/signup").permitAll()
                        .requestMatchers("/auth/**", "auth-logout").authenticated()
                        .anyRequest().authenticated()
                )
                .addFilterBefore(new JWTFilter(jwtUtil, memberRepository, tokenBlacklistService), UsernamePasswordAuthenticationFilter.class)
                .sessionManagement((session) -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
                .cors((cors) -> cors.configurationSource(new CorsConfigurationSource() {
                    @Override
                    public CorsConfiguration getCorsConfiguration(HttpServletRequest request) {
                        CorsConfiguration configuration = new CorsConfiguration();
                        configuration.setAllowedOrigins(Collections.singletonList("http://localhost:3000"));
                        configuration.setAllowedMethods(Collections.singletonList("*"));
                        configuration.setAllowCredentials(true);
                        configuration.setAllowedHeaders(Collections.singletonList("*"));
                        configuration.setMaxAge(3600L);
                        configuration.setExposedHeaders(Collections.singletonList("Authorization"));
                        return configuration;
                    }
                }));

        return http.build();
    }
}

File: jwtKotlin/src/main/kotlin/com/member/jwtkotlin/config/SecurityConfig.kt
package com.member.jwtkotlin.config

import com.member.jwtkotlin.jwt.JWTFilter
import com.member.jwtkotlin.jwt.JWTUtil
import com.member.jwtkotlin.repository.MemberRepository
import com.member.jwtkotlin.repository.RefreshRepository
import com.member.jwtkotlin.service.CustomMemberDetailService
import com.member.jwtkotlin.service.TokenBlacklistService
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.ProviderManager
import org.springframework.security.authentication.dao.DaoAuthenticationProvider
import org.springframework.security.config.Customizer
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configurers.*
import org.springframework.security.config.annotation.web.configurers.AuthorizeHttpRequestsConfigurer.AuthorizationManagerRequestMatcherRegistry
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.web.cors.CorsConfiguration

@Configuration
@EnableWebSecurity
class SecurityConfig(
    private val authenticationConfiguration: AuthenticationConfiguration,
    private val jwtUtil: JWTUtil,
    private val refreshRepository: RefreshRepository,
    private val customMemberDetailService: CustomMemberDetailService,
    private val tokenBlacklistService: TokenBlacklistService
) {
    @Bean
    fun bCryptPasswordEncoder(): BCryptPasswordEncoder {
        return BCryptPasswordEncoder()
    }

    @Bean
    fun authenticationProvider(): DaoAuthenticationProvider {
        val authProvider = DaoAuthenticationProvider()
        authProvider.setUserDetailsService(customMemberDetailService)
        authProvider.setPasswordEncoder(bCryptPasswordEncoder())
        return authProvider
    }

    @Bean
    @Throws(Exception::class)
    fun authenticationManager(): AuthenticationManager {
        val authenticationManager = authenticationConfiguration.authenticationManager
        (authenticationManager as ProviderManager).providers.add(authenticationProvider())
        return authenticationManager
    }

    @Bean
    @Throws(Exception::class)
    fun securityFilterChain(
        http: HttpSecurity,
        refreshRepository: RefreshRepository?,
        memberRepository: MemberRepository?
    ): SecurityFilterChain {
        http
            .csrf { csrf: CsrfConfigurer<HttpSecurity> -> csrf.disable() }
            .formLogin { form: FormLoginConfigurer<HttpSecurity> -> form.disable() }
            .httpBasic { basic: HttpBasicConfigurer<HttpSecurity> -> basic.disable() }
            .authorizeHttpRequests { auth ->
                auth
                    .requestMatchers("/login", "/", "/signup").permitAll()
                    .requestMatchers("/auth/**", "auth-logout").authenticated()
                    .anyRequest().authenticated()
            }

            .addFilterBefore(
                JWTFilter(jwtUtil, memberRepository!!, tokenBlacklistService),
                UsernamePasswordAuthenticationFilter::class.java
            )
            .sessionManagement { session: SessionManagementConfigurer<HttpSecurity?> ->
                session.sessionCreationPolicy(
                    SessionCreationPolicy.STATELESS
                )
            }
            .cors { cors: CorsConfigurer<HttpSecurity?> ->
                cors.configurationSource {
                    val configuration = CorsConfiguration()
                    configuration.allowedOrigins = listOf("http://localhost:3000")
                    configuration.allowedMethods = listOf("*")
                    configuration.allowCredentials = true
                    configuration.allowedHeaders = listOf("*")
                    configuration.maxAge = 3600L
                    configuration.exposedHeaders = listOf("Authorization")
                    configuration
                }
            }

        return http.build()
    }
}